src/android/nl/xservices/plugins/accessor/AbstractCalendarAccessor.java

```diff
@@ -384,11 +393,44 @@ private Map<String, Event> fetchEventsAsMap(Event[] instances, String calendarId
                 event.title = cursor.getString(cols[3]);
                 event.startDate = cursor.getString(cols[4]);
                 event.endDate = cursor.getString(cols[5]);
+                event.duration = cursor.getString(cols[6]);
+
+                // If a duration is provided but not the end date, calculate the end date from the duration
+                if (event.endDate == null && event.duration != null) {
+                    try {
+                        // split into period and duration
+                        String[] v = event.duration.split("T");
+                        Period p;
+                        Duration d;
+                        if (v.length == 1) { // has only date-based fields (or is a bad format)
+                            if (event.duration.contains("S") || event.duration.contains("M") || event.duration.contains("H")) {
+                                // Assume time duration given as a period (although M could be for months, it's less likely)
+                                p = Period.ZERO;
+                                d = Duration.parse("PT" + event.duration.substring(1));
+                            } else {
+                                p = Period.parse(event.duration);
+                                d = Duration.ZERO;
+                            }
+                        } else {
+                            if ("P".equals(v[0])) { // only has time-based fields
+                                p = Period.ZERO;
+                            } else {
+                                p = Period.parse(v[0]);
+                            }
+                            d = Duration.parse("PT" + v[1]);
+                        }
+                        Instant i = new Date(Long.parseLong(event.startDate)).toInstant();
+                        i = i.plus(p).plus(d);
+                        event.endDate = Long.toString(i.toEpochMilli());
+                    } catch (Exception e) {
+                        Log.d(LOG_TAG, "Couldn't parse duration: " + event.duration);
```

Comment:
Should we somehow ignore this event in this case? Not certain how to do that - returning `null` I think.

I suppose it won't be shown anyway, but might as well not process it on the app in that case

Reply:
Nah, because the app will receive it with what it gets given (so everything except the end-date) and then our check on the app will discard it as an invalid event, but will log it, so we can see it in the logs - which is useful if any users complain they aren't seeing events